libs/providers/vmware.py

```
@@ -330,6 +330,39 @@ def stop_vm(self, vm):
+            vm.ShutdownGuest()
+            LOGGER.info(f"Requested graceful shutdown for VM {vm.name}, waiting up to {timeout}s")
+            for sample in TimeoutSampler(
+                wait_timeout=timeout,
+                sleep=5,
+                func=lambda: vm.runtime.powerState,
+            ):
+                if sample != vm.runtime.powerState.poweredOn:
+                    LOGGER.info(f"VM {vm.name} gracefully shut down")
+                    return
+        except (vim.fault.ToolsUnavailable, vim.fault.InvalidPowerState):
+            LOGGER.warning(f"VMware Tools unavailable on VM {vm.name}, falling back to hard PowerOff")
```

Ruff (0.15.9)

[warning] Logging statement uses f-string

(G004)